### PR TITLE
List Index Expression

### PR DESCRIPTION
**Pull request by AlmanaX-21**
List index expression: Adds expression to return the index(indices) of object from a list.

---
**Targeted Minecraft version: Any**
**Requirements:No requirements**
**Related Issues: https://github.com/SkriptLang/Skript/issues/3933**

### CODE REVIEW

**Comment by AlmanaX-21** [comment]
I am not sure why some changes occurred in skript aliases, those weren't done manually.

**Comment by TheLimeGlass** [comment]
I had a note to add `the (last|first) %number% elements out of` and this pull fits it, if you're still looking for ideas.

File: src/main/java/ch/njol/skript/expressions/ExprListIndex.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+			"[all] ind(ex|ic)es of %object% in %objects%");
+	}
+
+	private Expression<?> elementExpr, listExpr;
+
+	private boolean variable, all, first;
+
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, SkriptParser.ParseResult parseResult) {
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprListIndex.java:64]: Rather than importing `SkriptParser.ParseResult` just import `ParseResult`

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprListIndex.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+@Description("Returns the index of an object from a list.")
+@Examples({
+	"command /listindex:",
+	"\ttrigger:",
+	"\t\tset {_list::*} to 1,2,3,4,5,1",
+	"\t\tsend \"%first index of 1 in %{_list::*}%%\"",
+	"\t\tsend \"%last index of 1 in %{_list::*}%%\"",
+	"\t\tsend \"%all indexes of 1 in %{_list::*}%%\""
+})
+public class ExprListIndex extends SimpleExpression<Object> {
```
[the remark this answers is not in the document]
Comment (AlmanaX-21) [reply src/main/java/ch/njol/skript/expressions/ExprListIndex.java:51]: Not sure but how would I know when you guys merge this, this isn't version dependant either.

File: src/main/java/ch/njol/skript/expressions/ExprListIndex.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	private Expression<?> elementExpr, listExpr;
+
+	private boolean variable, all, first;
+
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		variable = (parseResult.mark & 1) == 1;
```
[the remark this answers is not in the document]
Comment (AlmanaX-21) [reply src/main/java/ch/njol/skript/expressions/ExprListIndex.java:65]: I was gonna remove this just forgot to do that will fix all these variable usages

File: src/main/java/ch/njol/skript/expressions/ExprListIndex.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprListIndex.java:51]: > Not sure but how would I know when you guys merge this, this isn't version dependant either.

Generally I believe it's done like `@Since("INSERT VERSION")`, and the version is added when it's merged.

File: src/main/java/ch/njol/skript/expressions/ExprListIndex.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	protected
+	Object[] get(Event e) {
+		Object element = elementExpr.getSingle(e);
+		if (element == null)
+			return null;
+		if (variable) {
+			Object o = ((Variable<?>) listExpr).getRaw(e);
+			if (o instanceof TreeMap<?, ?>) {
+				TreeMap<?, ?> listVar = new TreeMap<>();
+				@SuppressWarnings("unchecked")
```
[the remark this answers is not in the document]
Comment (AlmanaX-21) [reply src/main/java/ch/njol/skript/expressions/ExprListIndex.java:89]: *uhm* that seems a little extra but I'll change that ig

File: skript-aliases
Comment (AyhamAl-Ali) [inline skript-aliases:1]: Make sure to revert this change

File: src/main/java/ch/njol/skript/expressions/ExprListIndex.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	"\ttrigger:",
+	"\t\tset {_list::*} to 1,2,3,4,5,1",
+	"\t\tsend \"%first index of 1 in %{_list::*}%%\"",
+	"\t\tsend \"%last index of 1 in %{_list::*}%%\"",
+	"\t\tsend \"%all indexes of 1 in %{_list::*}%%\""
+})
+public class ExprListIndex extends SimpleExpression<Object> {
+	static {
+		Skript.registerExpression(ExprListIndex.class, Object.class, ExpressionType.COMBINED,
+			"[(first|last)] index of %object% in %objects%",
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprListIndex.java:54]:
```suggestion
			"[first|last] index of %object% in %objects%",
```

File: src/main/java/ch/njol/skript/expressions/ExprListIndex.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+			"[all] ind(ex|ic)es of %object% in %objects%");
+	}
+
+	private Expression<?> elementExpr, listExpr;
+
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprListIndex.java:59]:
```suggestion
```

File: src/main/java/ch/njol/skript/expressions/ExprListIndex.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+		if (variable && !(listExpr instanceof Variable)) {
+			Skript.error(listExpr + " is not a list variable");
+			return false;
+		}
+		return LiteralUtils.canInitSafely(elementExpr, listExpr);
+	}
+
+	@Override
+	@Nullable
+	protected Object[] get(Event e) {
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprListIndex.java:81]: Change all `e` to `event`
```suggestion
	protected Object[] get(Event event) {
```

File: src/main/java/ch/njol/skript/expressions/ExprListIndex.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	@Override
+	public boolean isSingle() {
+		return !all;
+	}
+
+	@Override
+	public Class<?> getReturnType() {
+		return variable ? String.class : Long.class;
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprListIndex.java:112]: Shouldn't this be `Long.class` always?

**Review by AyhamAl-Ali — CHANGES_REQUESTED** [review]
On-the-fly review, nice work 👏
